Resolved potential GC borrow hazards in ModuleTree::append_handler, ModuleTree::append_dynamic_module_handler, and ModuleTree::fetch_module_descendants'

chickenleaf opened the pull request:
**Files:** `components\script\script_module.rs`

**PR Description**

**Issue:**
Addressed GC borrow hazards in the following functions:
  1.   `ModuleTree::append_handler`
  2.   `ModuleTree::append_dynamic_module_handler`
  3.   `ModuleTree::fetch_module_descendants`

These functions previously performed mutable borrows that could lead to a GC panic under certain conditions.

**Changes:**


**Impact:**
1. Prevents GC panics by avoiding mutable borrows during operations that might trigger garbage collection.
2. Improves the stability of the module fetching and handling process.

**Verification:**
1. Verified by running the `./mach test-wpt tests/wpt/tests/html/semantics/scripting-1/the-script-element/module` suite.
2. Followed the steps to reproduce the original GC panic, no crashes, confirming the issue is resolved.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #33926 and #33927 

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

jdm commented on components/script/script_module.rs:
```diff
@@ -357,14 +357,14 @@ impl ModuleTree {
 
+        let new_promise = Promise::new_in_current_realm(comp);
+
         let mut promise = self.promise.borrow_mut();
```
This borrow is still active when we call append_native_handler, so we need to reduce its scope.

jdm replied:
Ideally we only use borrow calls in expressions/statements that need them, rather than open-ended ones like this.

jdm commented on components/script/script_module.rs:
```diff
@@ -393,11 +393,12 @@ impl ModuleTree {
 
+        let new_promise = Promise::new_in_current_realm(comp);
+
         let mut promise = self.promise.borrow_mut();
```
Same issue.

chickenleaf committed:
GC borrow hazard ModuleTree fix

Signed-off-by: L Ashwin B <lashwinib@gmail.com>

chickenleaf committed:
Merge branch 'main' into GC-Borrow-Hazard-ModuleTree

Loirooriol committed:
Prevent `stretch` from producing a negative size (#33951)

Signed-off-by: Oriol Brufau <obrufau@igalia.com>

jdm commented on components/script/script_module.rs:
```diff
@@ -758,20 +759,21 @@ impl ModuleTree {
-                    assert!(visited_urls.get(&url).is_some());
-
-                    let options = options.descendant_fetch_options();
+                    debug_assert!(self.visited_urls.borrow().contains(&url));
```
```suggestion
                    assert!(self.visited_urls.borrow().contains(&url));
```

chickenleaf replied:
Are we using assert for all places that has a potential GC hazard? Even after making changes to ensure safety?

jdm replied:
The original code used assert, so we shouldn't modify it in this PR. The specification this code is implementing explicitly has an assert step.

jdm reviewed (approved):
Looking good!

chickenleaf committed:
reverted debug_assert back to assert

Signed-off-by: L Ashwin B <lashwinib@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

chickenleaf committed:
merge conflicts 1.0

Signed-off-by: L Ashwin B <lashwinib@gmail.com>

jdm commented:
You'll need to address the build errors caused by the changes in #33937.

chickenleaf committed:
merging and updating the branch

merge branch 'main' of https://github.com/chickenleaf/servo into GC-Borrow-Hazard-ModuleTree

chickenleaf committed:
added can_gc arguments to the function calls

Signed-off-by: L Ashwin B <lashwinib@gmail.com>

chickenleaf commented:
> You'll need to address the build errors caused by the changes in #33937.

done

Loirooriol commented:
It's confusing to have the commit from https://github.com/servo/servo/pull/33951 here because it's totally unrelated,  I guess you updated your PR doing a merge instead of a rebase or something like that.
